### DISCUSSION TITLE
What does MAUI offer to easen the process of drawing custom controls

### DISCUSSION THREAD

**Discussion by mshwf** [dotnet/maui#102] [General]
This is probably one of the most needed changes in MAUI.

I don't want to be bound to the platform-only presence.
1. I wish MAUI provides all controls of each platform's UI language (Material for Android, Cupertino for iOS.. other platforms').
3. The potential for each control to be fully customizable, (may be, something like `ControlTemplate` in WPF?).
2. Be able to create custom controls, with a variety of prime controls that can be customized altogether by shape, color (and gradients) animation.. etc.

Will MAUI offer that experience? how?
Skia is a bit overhead and requires a learning curve by its own, Flutter has had an amazing job regardig this point> using Skia as part of the framework itself.

What is MAUI's key solution to this?

**Comment by awulkan** [comment]
This is one of the most important bits in my opinion. I haven't used Xamarin, but I've used Flutter. If MAUI can't compete with Flutter in terms of customization, then it has already lost a huge part of the market.

**Comment by jingsong-liu** [comment]
I have used WPF xaml, I agree this is the most needed changes in MAUI because it's really not a good experience in WPF to create customized controls

**Comment by davidortinau** [comment]
> What is MAUI's key solution to this?

We don't have an official one yet, so let's consider some options. 

Would you provide some concrete examples from previous projects and perhaps some psuedo-code spec of what you might like it to look like?

Is it on a control by control basis you would need to opt for "customized altogether by shape, color (and gradients) animation.. etc."?

Can you provide some examples of control designs that cannot be achieved easily today with native controls? 

Which controls do you most often need/want to opt into drawing for?

Related:
* https://github.com/xamarin/Xamarin.Forms/pull/9218
* https://github.com/xamarin/Xamarin.Forms/pull/9264
* we have a PR incoming for clipping/masking

**Comment by mshwf** [comment]
@davidortinau You don't have plans yet, but already set the same old approach> platform renderers, even MAUI is just a fork of Xamarin.Forms!
this already cut the road to new ideas, at the best we would think of just small enhancements.

I would dream of a UI framework like Flutter in .NET. I'm .NET developer I want to stick with the .NET stack as long as it's evolving and improving, but I wasn't satisfied with the solution that Xamarin offers, however, you were doing amazing job and improving it with each release (v. 4.7 is great to offer shapes and gradients), but the problem in the Xamarin approach itself, that is tightly coupled with the underlying platform UI, I know you are changing that, like introducing RadioButton and Expander controls, but it's not clear the vision you're taking, is it a platform only controls, or you are open to custom controls? in that case, you may need to impove the Xamarin kit to make this vision approachable, not only for library authors but for cross-platform devs who don't have any native experience, and don't want to have.

I tried Flutter and I can say it was most pleasant experience for me and for many others (look at the recent Stackoverflow surveys), also a very welcoming framework for beginners, why?: the UI kit is huge and productivity tools like, the inspector and hot reload (Xamarin has that), and the newcomer plugin: Adobe XD to Flutter.

I don't know if something like that is possible to implement in MAUI with renderers.

If we go with the .NET familiar approach, then I want the same WPF experience, where every control has its template that can be customized, and with XAML support for animation and storyboards. but again being stuck with every platform renderer will eventually lead to a lot of bugs!

hope that answers your question,
Sorry for being lengthy, good luck!

**Comment by awulkan** [comment]
@mshwf I agree. I think a lot of people don't really want an evolution of Xamarin, they want something totally new like Flutter.

Releasing yet another GUI framework, and bringing the old stuff from Xamarin with it, is not a good idea imo. Microsoft will have to support MAUI for many years, so why not create something that will actually be attractive to developers during that time? It feels like Flutter has already "won" before the development of MAUI has even really begun.

Flutter was able to get developers to learn a whole new language just to use the framework. A language that was almost completely dead. Imagine if Microsoft created an equally attractive framework, but with C#/F#. It would be a lot more enticing for developers than Flutter with Dart will ever be.

**Comment by thekeviv** [comment]
@davidortinau As my two cents, what I like in flutter is the composability model due in part to many of their "mid-level" (not high level and not low level) widgets. They provide very nice basic components (widgets) which can be **composed** together to create very functional and good looking apps. I think Xamarin Forms focused quite heavily on providing many pre-built high level controls. Without trying to underplay the importance of higher level controls, I think that while they are great for Enterprise apps, I almost always find that the UI's for the apps I want to build tend to be highly customized versions of these controls. For that, the slope of learning how to customize these controls was always higher in my opinion in Xamarin as compared to other platforms. I think that the platform tried to go in that route at some time but that actually ended up bloating the API. I would like more control of the pixels on my screen with an easier to use cross platform API. Something similar to how layers work in photo editing apps but with more control and a cross platform implementation.

In any case, I love Xamarin and can't wait to get on the MAUI train!
